### PR TITLE
chore: update min version of sentry-sdk

### LINKED ISSUE #8361: [Bug]: Deprecation warning for sentry_sdk.Hub
### Describe the bug

Hi! 

Deprecation warning detailed below.

```
/venv/lib/python3.11/site-packages/wandb/analytics/sentry.py:90: SentryHubDeprecationWarning: `sentry_sdk.Hub` is deprecated and will be removed in a future major release. Please consult our 1.x to 2.x migration guide for details on how to migrate `Hub` usage to the new API: https://docs.sentry.io/platforms/python/migration/1.x-to-2.x
    self.hub = sentry_sdk.Hub(client)
```

Thank you!! 🙏

### PR DESCRIPTION
**Pull request by jacobromero**
Description
-----------

- Fixes WB-20890
- Fixes #8361

This PR bumps the minimum required version of `sentry-sdk` package to 2.0.0 and updates the sentry logging code to account for the removal of the `Hub` class.

- [x] I updated CHANGELOG.md, or it's not applicable

Testing
-------
How was this PR tested?

- Unit tests
- Manually raising exception and verify output is expected in sentry webpage compared to current logs.
- Manually logged a message and verify output is expected in sentry webpage compared to current logs.

#### Testing no data leak between sentry scopes
Created a script with `sentry_sdk.init(...)` and `wandb.init(...)` to setup multiple scopes for sentry_sdk. 
Added a tag to the scope created by `sentry_sdk.init`
Added calls to `sentry_sdk.capture_message/error` and `wandb._sentry.message/exception` simulating writing to two different sentry projects (both with different DSNs), 

- Test cases were:
  - `wandb.init` -> `sentry_sdk.init`
    - `wandb._sentry.message/exception` -> `sentry_sdk.capture_message/error` -> Output: 1 message to each DSN, no tags were shared between the messages
    -  `sentry_sdk.capture_message/error` -> `wandb._sentry.message/exception` -> Output: 1 message to each DSN, no tags were shared between the messages
  - `sentry_sdk.init` -> `wandb.init`
    - `wandb._sentry.message/exception` -> `sentry_sdk.capture_message/error` -> Output: 1 message to each DSN, no tags were shared between the messages
    -  `sentry_sdk.capture_message/error` before `wandb._sentry.message/exception` -> Output: 1 message to each DSN, no tags were shared between the messages
  - `sentry_sdk.init` -> `sentry_sdk.capture_message/error` -> `wandb._sentry.init` -> `wandb.capture_message/exception` -> Output: 1 message to each DSN, no tags were shared between the messages
  -  `wandb._sentry.init` -> `wandb.capture_message/exception` -> `sentry_sdk.init` -> `sentry_sdk.capture_message/error` -> Output: 1 message to each DSN, no tags were shared between the messages

### CODE REVIEW

**Comment by dmitryduev** [comment]
@jacobromero the main scenario we need to test is the following:
- set up a script that imports sentry and points it at a sentry dsn
- set up out dsn to point at say sdk-client (we're not using that project yet, so you'll only see your events there)
- import wandb in different places in the script (before and after setting up the script's sentry)

the reason why we've been using their private api's was bc the default set up steps were resulting in us picking up third-party events (or vice versa), so we need to verify it does not happen, and the events go to their respective projects/dsns.

**Review by kptkin — APPROVED** [review]

File: CHANGELOG.md
Comment (kptkin) [inline CHANGELOG.md:47]: you are updating a release section as well as changing another entry.
it should go to the unreleased section

File: tests/unit_tests/test_analytics/test_sentry.py
Comment (kptkin) [inline tests/unit_tests/test_analytics/test_sentry.py:32]: this is not ideal, that we could bind function just on import... could you potentially move it closer to when it is actually used?

File: tests/unit_tests/test_analytics/test_sentry.py
Comment (kptkin) [inline tests/unit_tests/test_analytics/test_sentry.py:41]: could we patch it instead of setting it up in the top of the file?

File: tests/unit_tests/test_analytics/test_sentry.py
Comment (kptkin) [inline tests/unit_tests/test_analytics/test_sentry.py:42]: same as above

File: CHANGELOG.md
Comment (jacobromero) [reply CHANGELOG.md:47]: Oops this was due to pulling in the latest changes, but missed updating this

File: tests/unit_tests/test_analytics/test_sentry.py
Comment (jacobromero) [reply tests/unit_tests/test_analytics/test_sentry.py]: Good point, move this to a pytest fixture to avoid needing to spin up an flask server for each test. Which also avoids binding a port on import.

File: tests/unit_tests/test_analytics/test_sentry.py
Comment (kptkin) [inline tests/unit_tests/test_analytics/test_sentry.py:48]: i would think it is better to just patch it directly in the tests, this is quite a big side-effect of turning the error reporting on

File: tests/unit_tests/test_analytics/test_sentry.py
Comment (kptkin) [inline tests/unit_tests/test_analytics/test_sentry.py:300]: should just move the docstring in the relevant test

File: tests/unit_tests/test_analytics/test_sentry.py
Comment (kptkin) [inline tests/unit_tests/test_analytics/test_sentry.py:131]: this sleep is really not ideal for unit testing (or in general in testing)
i'm pretty sure it will cause quite a bit of flake

Comment (jacobromero) [reply tests/unit_tests/test_analytics/test_sentry.py:131]: I agree, the issue is that sentry_sdk appears to be sending events asynchronously. Without waiting for some time the tests consistently fail. 

This method mitigates flakiness by having a default 5 second timeout. Because the events are sent over localhost. There is no real latency, just waiting for sentry_sdk to actually send the events.

Comment (kptkin) [reply tests/unit_tests/test_analytics/test_sentry.py:124]: yeah, i get the reasoning behind it, just thinking maybe this approach is not optimal. 
did you check how sentry does it testing internally? ideally if we could mock a bunch of things and assert on that, that would be preferable

Comment (jacobromero) [reply tests/unit_tests/test_analytics/test_sentry.py:124]: An option is to implement our own [transport class](https://github.com/getsentry/sentry-python/blob/master/tests/test_monitor.py#L23-L26) and capture the events as they are called from there.

We would then need to add the ability to provide a transport class to the wandb Sentry class.

File: tests/unit_tests/test_analytics/test_sentry.py
Comment (kptkin) [inline tests/unit_tests/test_analytics/test_sentry.py:154]:
```suggestion
def test_wandb_sentry_init_after_client_init(relay):
```

File: tests/unit_tests/test_analytics/test_sentry.py
Comment (kptkin) [inline tests/unit_tests/test_analytics/test_sentry.py:151]: seems redundant for a one-line

File: tests/unit_tests/test_analytics/test_sentry.py
Comment (kptkin) [reply tests/unit_tests/test_analytics/test_sentry.py:124]: Thanks, i will need to think about it.

File: tests/unit_tests/test_analytics/test_sentry.py
Comment (kptkin) [inline tests/unit_tests/test_analytics/test_sentry.py:171]:
```suggestion
    with mock.patch.dict(os.environ, {
        wandb.env.ERROR_REPORTING: 'true',
        wandb.env.SENTRY_DSN: WANDB_DSN_URL_FORMAT.format(port=relay.port)
    }):
        wandb_sentry = wandb.analytics.Sentry()
        wandb_sentry.setup()
```

File: tests/unit_tests/test_analytics/test_sentry.py
Comment (kptkin) [inline tests/unit_tests/test_analytics/test_sentry.py:21]: i think it is better to just add it in the test, as it is just complicates reading

File: tests/unit_tests/test_analytics/test_sentry.py
Comment (kptkin) [inline tests/unit_tests/test_analytics/test_sentry.py:69]: would be nice to finish the thread

File: tests/unit_tests/test_analytics/test_sentry.py
Comment (kptkin) [inline tests/unit_tests/test_analytics/test_sentry.py:128]: maybe instead of this function, let's add some logic around envelope (maybe another class?) that will return the relevant information, like: `public_key`, `payload`, `tags`, `project_id`

File: tests/unit_tests/test_analytics/test_sentry.py
Comment (kptkin) [inline tests/unit_tests/test_analytics/test_sentry.py:27]: now, that i look at this what do you think about only keeping:
```
SENTRY_DSN_FORMAT = "http://{key}@{address}/{project}"
```
i know there is some repetition, but i think this is the only interesting part of all these vars

File: tests/unit_tests/test_analytics/test_sentry.py
Comment (kptkin) [inline tests/unit_tests/test_analytics/test_sentry.py:79]:
```suggestion
    other_sentry_client_message = "client message"
    wandb_sentry_client_message = "wandb message"
```

Comment (kptkin) [reply tests/unit_tests/test_analytics/test_sentry.py:79]: basically have a naming that explicitly distinguish between "wandb" sentry client vs other (non-wandb) sentry client

File: tests/unit_tests/test_analytics/sentry_relay.py
Comment (kptkin) [inline tests/unit_tests/test_analytics/sentry_relay.py:84]: couldn't find where this is being used

File: tests/unit_tests/test_analytics/test_sentry.py
Comment (jacobromero) [reply tests/unit_tests/test_analytics/test_sentry.py:27]: personally I think it is best to keep these defined as constants in the file. My rationale is

- If in the future more tests are added, reuse of these are easier, and avoids possibly mistyping.
- If for any reason we need to change one of the values, it can be changed from a single place

I think if we do want to remove the clutter from this file, we can move this to something like `constants.py` file or similar to avoid having too many things in this test file. Thoughts?

Comment (kptkin) [reply tests/unit_tests/test_analytics/test_sentry.py]: actually the opposite i would rather it be closer to the test and if there is a typo shouldn't it be caught in the test directly? these are all just values you use for the test, no special meaning as far as i can see

File: tests/unit_tests/test_analytics/constants.py
Comment (kptkin) [inline tests/unit_tests/test_analytics/constants.py:1]: i don't think it helps with readability, let's just have these closer to the test code, these are not expected be used by any other tests...

File: CHANGELOG.md
Comment (kptkin) [inline CHANGELOG.md:31]: could you move this to the unreleased section

File: tests/unit_tests/test_analytics/sentry_relay.py
Comment (kptkin) [inline tests/unit_tests/test_analytics/sentry_relay.py:26]: doc string here would be nice

File: tests/unit_tests/test_analytics/test_sentry.py
Comment (kptkin) [inline tests/unit_tests/test_analytics/test_sentry.py:95]: this one feels a bit like a hack could you think of a way to make it better?

File: tests/unit_tests/test_analytics/test_sentry.py
Comment (kptkin) [inline tests/unit_tests/test_analytics/test_sentry.py:108]:
```suggestion
    """
    This test ensures proper isolation between the Sentry instances initialized by another client and by wandb's client. 
    
    The test involves initializing the Sentry client first using sentry_sdk.init(), followed by initializing the wandb Sentry client. Both instances set distinct session tags.
 The test ensures that events from both the other client and wandb are successfully received by their servers, while confirming that the data remains isolated between the two Sentry instances. 
```

something like this describe the essence of the test, think of what would be useful for the other devs

File: tests/unit_tests/test_analytics/test_sentry.py
Comment (kptkin) [inline tests/unit_tests/test_analytics/test_sentry.py:111]:
```suggestion
    other_sentry_message = "client message"
    other_sentry_public_key = "OTHER_SENTRY_PUBLIC_KEY"
    other_sentry_project_id = "654321"
```
let's keep the names consistent either add client to all or none

File: tests/unit_tests/test_analytics/test_sentry.py
Comment (kptkin) [inline tests/unit_tests/test_analytics/test_sentry.py:114]: can you convert this one to also to a response instance?
also maybe you could define some comparison operator or method, when doing asserting
i think it will be more readable

File: tests/unit_tests/test_analytics/test_sentry.py
Comment (kptkin) [reply tests/unit_tests/test_analytics/test_sentry.py:25]: still don't like this function :)

Comment (kptkin) [reply tests/unit_tests/test_analytics/test_sentry.py:25]: we could add methods/or operators on the response class

**Comment by kptkin** [comment]
for tests that you ported from the landfill folder could you delete them from that folder as well?

File: tests/unit_tests/test_analytics/test_sentry.py
Comment (kptkin) [inline tests/unit_tests/test_analytics/test_sentry.py:311]:
```suggestion
    expected_wandb_sentry_responses = [SentryResponse(
        message="wandb sentry message",
        project_id="123456",
        public_key="WANDB_SENTRY_PUBLIC_KEY",
    ), SentryResponse(
        message="wandb sentry message2",
        project_id="123456",
        public_key="WANDB_SENTRY_PUBLIC_KEY",
    ), 
    ]
```

File: tests/unit_tests/test_analytics/test_sentry.py
Comment (kptkin) [inline tests/unit_tests/test_analytics/test_sentry.py:349]: probably better to use the expect_wandb_sentry_response2.message

Comment (kptkin) [reply tests/unit_tests/test_analytics/test_sentry.py:349]: or the relevant entry in the array

File: wandb/analytics/sentry.py
Comment (kptkin) [inline wandb/analytics/sentry.py:1]: could add a few more tests for the missing coverage in this file?

Comment (kptkin) [reply wandb/analytics/sentry.py:1]: also for the deleted landfill tests, could you point leave a comment, how you test it instead and if you don't test it why. thanks

File: tests/unit_tests/test_analytics/sentry_relay.py
Comment (kptkin) [inline tests/unit_tests/test_analytics/sentry_relay.py:113]: a comment here would be useful

File: wandb/analytics/sentry.py
Comment (jacobromero) [reply wandb/analytics/sentry.py:1]: > also for the deleted landfill tests, could you point leave a comment, how you test it instead and if you don't test it why. thanks

Do you mean on the tests comment which of the landfill tests they replace?

Comment (kptkin) [reply wandb/analytics/sentry.py:1]: sorry i meant the other way around, in the landfill test... just comment on the files

File: tests/unit_tests/test_analytics/sentry_relay.py
Comment (kptkin) [inline tests/unit_tests/test_analytics/sentry_relay.py:20]: why only tags are typed? haha

Comment (kptkin) [reply tests/unit_tests/test_analytics/sentry_relay.py:20]: also you could use:
`from __future__ import annotations` and use the newest typing style

File: tests/unit_tests/test_analytics/test_sentry.py
Comment (kptkin) [inline tests/unit_tests/test_analytics/test_sentry.py:72]: probably don't need these as are only used once

File: tests/unit_tests/test_analytics/test_sentry.py
Comment (kptkin) [inline tests/unit_tests/test_analytics/test_sentry.py:34]: probably don't need these as are only used once,
if you need to make a distinction, could just leave a comment

File: tests/unit_tests/test_analytics/sentry_relay.py
Comment (kptkin) [inline tests/unit_tests/test_analytics/sentry_relay.py:50]: why do you need this special class for adding the stack trace? it seems that it is enough to have the original one and just not set the stack trace argument

Comment (jacobromero) [reply tests/unit_tests/test_analytics/sentry_relay.py]: My thoughts were that it feels like its starting to overload the SentryResponse class to much.
By splitting into the exception class the purpose of the class is more clear and it would allow more flexibility if additional methods, fields, or verification methods were needed later.

But we can do this later if needed.

File: tests/unit_tests/test_analytics/sentry_relay.py
Comment (kptkin) [inline tests/unit_tests/test_analytics/sentry_relay.py:127]: why do you need this one? if you have if-else after that?

File: tests/unit_tests/test_analytics/sentry_relay.py
Comment (kptkin) [inline tests/unit_tests/test_analytics/sentry_relay.py:126]: should we just move it to else?

File: tests/unit_tests/test_analytics/test_sentry.py
Comment (kptkin) [inline tests/unit_tests/test_analytics/test_sentry.py:67]: sorry i meant just leave a comment on the sentry file in github
(see the example i left above)

File: landfill/functional_tests/debug/t4_sentry_tags.py
Comment (jacobromero) [inline landfill/functional_tests/debug/t4_sentry_tags.py:1]: `test_wandb_sentry_init_after_client_init` replaces the landfill test verifying wandb client adds tags to the sentry messages.

File: landfill/functional_tests/debug/t7_sentry_session.py
Comment (jacobromero) [inline landfill/functional_tests/debug/t7_sentry_session.py:1]: `test_wandb_sentry_init_after_client_init` replaces the landfill test verifying that events are sent to Sentry from the wandb sentry client.

File: landfill/functional_tests/debug/t8_sentry_session_no_finish.py
Comment (jacobromero) [inline landfill/functional_tests/debug/t8_sentry_session_no_finish.py:1]: `test_wandb_sentry_init_after_client_init` replaces the landfill test verifying that events are sent to Sentry from the wandb sentry client.

File: landfill/functional_tests/debug/t5_sentry_third_party.yea
Comment (jacobromero) [inline landfill/functional_tests/debug/t5_sentry_third_party.yea:1]: `test_wandb_sentry_exception` replaces this landfill test verifying that the stack trace is present in the sentry event.

File: landfill/functional_tests/debug/t3_sentry_userproc_unix.yea
Comment (jacobromero) [inline landfill/functional_tests/debug/t3_sentry_userproc_unix.yea:1]: `test_wandb_sentry_exception` replaces this landfill test verifying that the stack trace is present in the sentry event.

File: landfill/functional_tests/debug/t6_sentry_third_party_no_wandb_error_reporting.yea
Comment (kptkin) [inline landfill/functional_tests/debug/t6_sentry_third_party_no_wandb_error_reporting.yea:1]: why are you deleting this test if you didn't replace it? if you think this test is not needed just leave a comment here

File: landfill/functional_tests/debug/t3_sentry_userproc_win.yea
Comment (kptkin) [inline landfill/functional_tests/debug/t3_sentry_userproc_win.yea:1]: we actually run all the unit tests on windows in main see here: https://github.com/wandb/wandb/blob/main/.circleci/config.yml#L497-L508

File: landfill/functional_tests/debug/t6_sentry_third_party_no_wandb_error_reporting.yea
Comment (jacobromero) [reply landfill/functional_tests/debug/t6_sentry_third_party_no_wandb_error_reporting.yea:1]: This one was replaced with `test_wandb_error_reporting_disabled` unit test.

you had also commented above, so I did not add a comment myself.

```
test_wandb_error_reporting_disabled replaces the landfill test verifying disabling of wandb Sentry client via environment variable.
```

File: landfill/functional_tests/debug/t3_sentry_userproc_win.yea
Comment (jacobromero) [reply landfill/functional_tests/debug/t3_sentry_userproc_win.yea:1]: Hmm interesting. I dont see it actually running in the checks here https://app.circleci.com/pipelines/github/wandb/wandb/39079/workflows/172218e8-b348-4043-9c94-2811e8c223b8

am I missing something? Or it is possibly misconfigured?

Comment (jacobromero) [reply landfill/functional_tests/debug/t3_sentry_userproc_win.yea:1]: oh wait it looks like it only runs on certain branches. Got it!
In that case we can also remove the other test.

File: landfill/functional_tests/debug/t3_sentry_userproc_unix.yea
Comment (jacobromero) [inline landfill/functional_tests/debug/t3_sentry_userproc_unix.yea:1]: `test_wandb_sentry_exception` replaces this landfill test verifying that the stack trace is present in the sentry event for windows environments..

File: landfill/functional_tests/debug/t3_sentry_userproc_win.yea
Comment (kptkin) [reply landfill/functional_tests/debug/t3_sentry_userproc_win.yea]: if you want to test it before you merge you can temporarily mock with the config file

File: landfill/functional_tests/debug/t6_sentry_third_party_no_wandb_error_reporting.yea
Comment (kptkin) [reply landfill/functional_tests/debug/t6_sentry_third_party_no_wandb_error_reporting.yea:1]: ah right sorry, my bad...

**Review by kptkin — APPROVED** [review]
LGTM!!!

File: tests/unit_tests/test_analytics/test_sentry.py
Comment (kptkin) [inline tests/unit_tests/test_analytics/test_sentry.py:130]: why? i think you should do it when you initialize the expected response, don't reach for the internals?

Comment (jacobromero) [reply tests/unit_tests/test_analytics/test_sentry.py]: Thats fair, my initial thought was that some tags are auto added to sentry when calling `configure_scope` which at first I was more interested in testing that data (i.e. tags) between the 3p sentry client and the wandb sentry client is not duplicated.

now the tests have become more encompassing so it makes sense that we test that the tags we expect the `configure_scope` function to add are also present.

File: tests/unit_tests/test_analytics/sentry_relay.py
Comment (dmitryduev) [inline tests/unit_tests/test_analytics/sentry_relay.py:68]: Is this used anywhere?

**Review by dmitryduev — APPROVED** [review]
Thanks a lot @jacobromero, LGTM (modulo a couple nits)!

File: tests/unit_tests/test_analytics/sentry_relay.py
Comment (jacobromero) [reply tests/unit_tests/test_analytics/sentry_relay.py:67]: It was only used when verifying the server setup, but we can manually construct the string there 😃

**Pull request merged**
